introduce async capable methods on existing classes

ali-ince opened the pull request:
TcpSocketClient/SocketClient changes for async api

1. Changed ITcpSocketClient#DisconnectAsync to ITcpSocketClient#Disconnect since no async stuff inside.
2. Added timeOut parameter to ITcpSocketClient#ConnectAsync to better handle time outs across connection retries.
3. Modified TcpSocketClient / SocketClient and affected tests to honour these changes.

ali-ince committed:
TcpSocketClient/SocketClient changes for async api

1. Changed ITcpSocketClient#DisconnectAsync to ITcpSocketClient#Disconnect since no async stuff inside.
2. Added timeOut parameter to ITcpSocketClient#ConnectAsync to better handle time outs across connection retries.
3. Modified TcpSocketClient and SocketClient to honour these changes.

zhenlineo commented on Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs:
```diff
@@ -81,27 +82,32 @@ public TcpSocketClient(EncryptionManager encryptionManager, bool keepAlive, bool
-                    innerErrors = innerErrors == null ? new AggregateException(error) : new AggregateException(innerErrors, error);
+                    try
+                    {
+                        cancellationSource.Token.ThrowIfCancellationRequested();
```
If I understand this correctly, this timeout newly added is not for timing out `_client.ConnectAsync` but for this `for(var i = 0; i < address.Length; i++)`.

My question is if this timeout is necessary? Should we consider having timeout on `_client.ConnectAsync` instead?

ali-ince replied:
I have done that change to remove Wait() call on [this line](https://github.com/neo4j/neo4j-dotnet-driver/blob/1.5/Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs#L70) and move the timeout logic to a more manageable level - without blocking.

ali-ince replied:
Before making any changes to the current behaviour, we should clarify what we will do when a single socket connection times out in the for-loop. Do we cancel all the connection process (throw an exception) or will we continue trying the next IPAddress available? 

It depends on how "establishing a connection" is conceived on the developer side. Is it the maximum time allowed to get a working SocketConnection or is it the maximum time to connect to an IPAddress as a pure Socket - which on failure an actual working SocketConnection could take some multitude of the timeout value given.

Keeping it as it is for not introducing a behavioral change. Will work on it when we clarify its behavior-to-be.

ali-ince committed:
InputStream/OutputStream/PackStream changes for async api

1. Introduced async versions of existing methods to IInputStream/IChunkedInputStream.
changes.
2. Introduced IOutputStream#FlushAsync, IReader#ReadAsync, IWriter#FlushAsync and PackStreamMessageFormatV1.WriterV1#FlushAsync.
3. Introduced async versions of existing methods to PackStream.Unpacker,
PackStreamMessageFormatV1.ReaderV1.
4. Modified ChunkedInputStream/ChunkedOutputStream to honour these
5. Cloned existing tests to target new async methods.
